### DISCUSSION TITLE
clearTaskInstances API no longer returns cleared task ids

### DISCUSSION THREAD

**Discussion by ed-sparkes** [apache/airflow#28088] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Definitely present in Airflow 2.4.2

https://airflow.apache.org/api/v1/dags/{dag_id}/clearTaskInstances

is supposed to return a list of cleared task instances and it used to work as expected but some version between 2.3.4 and 2.4.2 the behaviour changed. 

in this file `https://github.com/apache/airflow/blob/main/airflow/api_connexion/endpoints/task_instance_endpoint.py`

it appears that something is mutating `task_instances` if not dry run. If i run dry run the task_instances are returned, as soon as a i run non dry_run the response is always an empty list. 

```
    task_instances = dag.clear(dry_run=True, dag_bag=get_airflow_app().dag_bag, task_ids=task_ids, **data)
    if not dry_run:
        clear_task_instances(
            task_instances.all(),
            session,
            dag=dag,
            dag_run_state=DagRunState.QUEUED if reset_dag_runs else False,
        )

    return task_instance_reference_collection_schema.dump(
        TaskInstanceReferenceCollection(task_instances=task_instances.all())
    )
```

I looked into `https://github.com/apache/airflow/blob/main/airflow/models/taskinstance.py` where the clear_task_instances function lives, i compared the code between 2.3.4 and 2.4.2 and the only difference i thought that may cause it is the addition of a `    session.flush()` at the end of the method. 

Any thoughts on this would be welcome as we cant upgrade and resolve some security dependencies until this is resolved as we rely on the response from this api. 

If the `session.flush()` is the culprit, i presume that was added for some other reason to resolve some other issue?

### What you think should happen instead

It should return the cleared tasks as per the documentation

### How to reproduce

Detail in the What happened section above

### Operating System

linux

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by tirkarthi** [comment]
I couldn't reproduce this on main branch with dry_run as false. There are also tests for this endpoint testing the response to return the list of task instances at https://github.com/apache/airflow/blob/518fd8479a1b9bfe303e184672231b6bc84687db/tests/api_connexion/endpoints/test_task_instance_endpoint.py#L891

```
http -a admin:admin POST :8080/api/v1/dags/example_branch_datetime_operator/clearTaskInstances 'task_ids:=["datetime_branch"]' only_failed:=false dry_run:=false
HTTP/1.1 200 OK
Connection: close
Content-Length: 690
Content-Type: application/json
Date: Sat, 26 Nov 2022 05:21:31 GMT
Server: Werkzeug/2.2.2 Python/3.10.6
X-Robots-Tag: noindex, nofollow

{
    "task_instances": [
        {
            "dag_id": "example_branch_datetime_operator",
            "dag_run_id": "manual__2022-11-26T04:48:39.482400+00:00",
            "execution_date": "2022-11-26T04:48:39.482400+00:00",
            "task_id": "datetime_branch"
        },
        {
            "dag_id": "example_branch_datetime_operator",
            "dag_run_id": "manual__2022-11-26T05:14:27.022615+00:00",
            "execution_date": "2022-11-26T05:14:27.022615+00:00",
            "task_id": "datetime_branch"
        },
        {
            "dag_id": "example_branch_datetime_operator",
            "dag_run_id": "scheduled__2022-11-25T00:00:00+00:00",
            "execution_date": "2022-11-25T00:00:00+00:00",
            "task_id": "datetime_branch"
        }
    ]
}
```

**Comment by potiuk** [comment]
Until better reproduction steps are provided - converting it into a discussion

**Comment by pso-naithanan** [comment]
Hi, 
I am able to reproduce the issue in our environment. We have noticed the issue when we want to re-run the failed task from the REST API(clearTaskInstances).he issue can be reproduce with Swagger API and from the python code.

The following is the scenario in which we are encountering the issue:    
    When trying to re-run the failed task from REST API()
    If we set the value of `only_failed` parameter to `True`, then `task_instances` value is null , but if we change the `only_failed` parameter to `False`, we are seeing the value for task_instances.

I am providing the following information :- 
    Screen shots for Swagger UI (First and Second images are for only_failed : `True` with null task_details. Third and Fourth images with only_failed : `False` and value of the task_details)
    Sample code to reproduce the issue.

<img width="1262" alt="clearTask_i" src="https://github.com/apache/airflow/assets/77467666/dc20dcd1-a449-40e2-9a23-c50dc4eae923">

<img width="1149" alt="clear_task_ii" src="https://github.com/apache/airflow/assets/77467666/73e2c86f-b1be-4cb9-a39a-61ed1bba2809">
<img width="1179" alt="clear_task_iii" src="https://github.com/apache/airflow/assets/77467666/fc805aa5-bafd-4eb9-b6cb-6168621c6981">
<img width="1173" alt="clear_task_iiii" src="https://github.com/apache/airflow/assets/77467666/6ce2cb24-e4d1-4557-ba23-6f07f872fefd">

`
import requests
from requests.auth import HTTPBasicAuth
auth=HTTPBasicAuth('test', 'ad%56tg#')


def clear_tasks():
    print("here inside clear_tasks")
    request_url = "http://localhost:8080/api/v1/dags/tutorial/clearTaskInstances"
    headers = {'Content-type': 'application/json','Accept':'application/json'}
    print(request_url)
    payload = {
        "dry_run": False,
        "task_ids": ["print_date"],
        "dag_run_id": "manual__2023-08-18T22:25:40.979789+00:00",
        "only_failed": False

    }

    response = requests.post(request_url,json=payload,headers=headers,auth=auth)
    if response.status_code == 403:
        raise requests.HTTPError(
            "You do not have a permission to perform this operation. "
            "Check Airflow RBAC roles for your account."
            f"{response.headers} / {response.text}"
        )
    elif response.status_code != 200:
        response.raise_for_status()
    else:
        return response.text


if __name__ == "__main__":
    response_text = clear_tasks()
    print(response_text)
`

**Reply by pso-naithanan** [reply]
@tirkarthi , i was able to reproduce the issue. Updated with my comments. Can you please move this as bug ?
